# SVM (Support Vector Machine)

- for classification using hyper-plane for Over-dimentional data

In [2]:
import pandas as pd

from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn import svm

dataset=pd.read_csv('train.csv')

dataset.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [3]:
le=preprocessing.LabelEncoder()
dataset["Sex"]=le.fit_transform(dataset["Sex"])

dataset["Embarked"]=le.fit_transform(dataset["Embarked"])

y=dataset['Pclass'] # Dependent variable

X=dataset.drop(['Pclass',"PassengerId","Name","Cabin","Ticket"],axis=1) # Independent variable

X.head()

Survived  Sex   Age  SibSp  Parch     Fare  Embarked
0         0    1  22.0      1      0   7.2500         2
1         1    0  38.0      1      0  71.2833         0
2         1    0  26.0      0      0   7.9250         2
3         1    0  35.0      1      0  53.1000         2
4         0    1  35.0      0      0   8.0500         2

# Train test split

In [4]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)
X_train.head()

Survived  Sex   Age  SibSp  Parch     Fare  Embarked
350         0    1  45.0      0      0  35.0000         2
124         1    1  12.0      1      0  11.2417         0
577         0    0  45.0      1      0  14.4583         0
422         0    0  28.0      1      1  14.4000         2
118         0    0   2.0      4      2  31.2750         2

In [5]:
X_test.head()

Survived  Sex   Age  SibSp  Parch      Fare  Embarked
14          0    0  14.0      0      0    7.8542         2
158         0    1  45.0      8      2   69.5500         2
762         1    0  36.0      1      2  120.0000         2
740         0    1  36.0      1      0   78.8500         2
482         1    0  63.0      0      0    9.5875         2

In [6]:
clf=svm.SVC(gamma=0.01,C=100)
clf.fit(X_train,y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [7]:
y_pred=clf.predict(X_test)
print('Accuracy for Pclass dependent var is....',accuracy_score(y_test,y_pred,normalize=True)*100,"%")

Accuracy for Pclass dependent var is.... 90.2621722846442 %


In [8]:
confusion_matrix(y_test,y_pred)

array([[ 64,   4,   2],
       [  4,  36,   9],
       [  1,   6, 141]], dtype=int64)

# For Survived as DV

In [9]:
y=dataset['Survived']
# Independent variable
X=dataset.drop(['Survived',"PassengerId","Name","Cabin","Ticket"],axis=1)

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

clf=svm.SVC(gamma=0.01,C=100)

clf.fit(X_train,y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [10]:
y_pred=clf.predict(X_test)
print('Accuracy for Survived dependent var is....',accuracy_score(y_test,y_pred,normalize=True)*100,"%")

Accuracy for Survived dependent var is.... 74.15730337078652 %


# Gender as DV

In [12]:
y=dataset['Sex']
# Independent variable
X=dataset.drop(['Sex',"PassengerId","Name","Cabin","Ticket"],axis=1)


X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

clf=svm.SVC(gamma=0.01,C=100)


clf.fit(X_train,y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [13]:
y_pred=clf.predict(X_test)
print('Accuracy for Gender dependent var is....',accuracy_score(y_test,y_pred,normalize=True)*100,"%")

Accuracy for Gender dependent var is.... 72.65917602996255 %


# Embarked as DV

In [14]:
y=dataset['Embarked']
# Independent variable
X=dataset.drop(['Embarked',"PassengerId","Name","Cabin","Ticket"],axis=1)

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

clf=svm.SVC(gamma=0.01,C=100)

clf.fit(X_train,y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [15]:
y_pred=clf.predict(X_test)
print('Accuracy for Embarked dependent var is....',accuracy_score(y_test,y_pred,normalize=True)*100,"%")

Accuracy for Embarked dependent var is.... 76.77902621722846 %


# Parch as DV

In [17]:
y=dataset['Parch']
# Independent variable
X=dataset.drop(['Parch',"PassengerId","Name","Cabin","Ticket"],axis=1)

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

clf=svm.SVC(gamma=0.01,C=100)


clf.fit(X_train,y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [18]:
y_pred=clf.predict(X_test)
print('Accuracy for Parch dependent var is....',accuracy_score(y_test,y_pred,normalize=True)*100,"%")

Accuracy for Parch dependent var is.... 79.7752808988764 %


# SibSp as Dv

In [20]:
y=dataset['SibSp']
# Independent variable
X=dataset.drop(['SibSp',"PassengerId","Name","Cabin","Ticket"],axis=1)

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)
clf=svm.SVC(gamma=0.01,C=100)
clf.fit(X_train,y_train)

SVC(C=100, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.01, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [21]:
y_pred=clf.predict(X_test)
print('Accuracy for SibSp dependent var is....',accuracy_score(y_test,y_pred,normalize=True)*100,"%")

Accuracy for SibSp dependent var is.... 74.90636704119851 %


# Inference

Accuracy for Pclass dependent var is...... 90.2621722846442 %, <br>
Accuracy for Survived dependent var is.... 74.15730337078652 % <br>
Accuracy for SibSp dependent var is....... 74.90636704119851 % <br>
Accuracy for Parch dependent var is.... ...79.7752808988764 % <br> 
Accuracy for Gender dependent var is...... 76.77902621722846 % <br> 
So, the Dependent variable to choose out of all the variables is the P-class variable as the dependent varible as it gives 90.26% <br>
Accuracy for the modelvwhen taken as the Dependent variable